rust/core-lib/src/tabs.rs

```diff
@@ -732,6 +743,55 @@ impl CoreState {
         }
     }
 
+    /// Handles changes in plugin files.
+    #[cfg(feature = "notify")]
+    fn handle_plugin_fs_event(&mut self, event: DebouncedEvent) {
+        use self::DebouncedEvent::*;
+        match event {
+            Create(ref path) | Write(ref path) => {
+                self.plugins.load_from_paths(&[path.clone()]);
+                if let Some(plugin) = self.plugins.get_from_path(path) {
+                    self.do_start_plugin(ViewId(0), &plugin.name);
+                }
+            }
+            // the way FSEvents on macOS work, we want to verify that this path
+            // has actually be removed before we do anything.
+            NoticeRemove(ref path) | Remove(ref path) if !path.exists() => {
+                if let Some(plugin) = self.plugins.get_from_path(path) {
+                    self.do_stop_plugin(ViewId(0), &plugin.name);
+                    self.plugins.remove_named(&plugin.name);
+                }
+            }
+            Rename(ref old, ref new) => {
+                if let Some(old_plugin) = self.plugins.get_from_path(old) {
+                    self.do_stop_plugin(ViewId(0), &old_plugin.name);
+                    self.plugins.remove_named(&old_plugin.name);
+                }
+
+                self.plugins.load_from_paths(&[new.clone()]);
+                if let Some(new_plugin) = self.plugins.get_from_path(new) {
+                    self.do_start_plugin(ViewId(0), &new_plugin.name);
+                }
+            }
+            Chmod(ref path) | Remove(ref path) => {
+                if let Some(plugin) = self.plugins.get_from_path(path) {
+                    self.do_stop_plugin(ViewId(0), &plugin.name);
+                    self.do_start_plugin(ViewId(0), &plugin.name);
+                }
+            }
+            _ => (),
+        }
```

Comment:
looking back at this big switch statement I do wonder if there's room for simplification, maybe by adding a few helper functions?

I also wonder if there might be a simplification possible: On any event, find the plugin (by folder) that has changed. If it's been removed, stop it; if it's been added, start it, and in all other cases restart it?

Just thinking out loud, this approach works it just feels a bit more fragile / easier to screw up. I'm not sure my idea is any good, though.